service/src/main/java/com/ge/predix/acs/service/policy/validation/PolicySetValidatorImpl.java

```diff
@@ -83,10 +87,70 @@ public void removeCachedConditions(final PolicySet policySet) {
     @Override
     public void validatePolicySet(final PolicySet policySet) {
         validateSchema(policySet);
+        List<String> obligationIds = validateObligationsAndGetIds(policySet.getObligations());
         for (Policy p : policySet.getPolicies()) {
             validatePolicyConditions(p.getConditions());
             validatePolicyActions(p);
+            vaLidatePolicyObligations(obligationIds, p);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private List<String> validateObligationsAndGetIds(final List<Obligation> obligations) {
+        List<String> ids = new ArrayList<>();
+        List<String> repeatedIds = new ArrayList<>();
+        int index = 0;
+
+        for (Obligation obligation : obligations) {
+            if (StringUtils.isEmpty(obligation.getId())) {
+                throw new PolicySetValidationException(
+                        String.format(PolicySetValidationMessages.EXCEP_OBL_ID_NULL_EMPTY, index));
+            }
+            if (ids.contains(obligation.getId())) {
+                repeatedIds.add(obligation.getId());
+            }
+
+            if (null == obligation.getActionTemplate()
+                    || ((Map<Object, Object>) obligation.getActionTemplate()).isEmpty()) {
+                throw new PolicySetValidationException(
+                        String.format(PolicySetValidationMessages.EXCEP_OBL_ACTION_TEMPLATE_NULL_EMPTY, index));
+            }
+            validateActionArguments(index, obligation.getActionArguments());
+            ids.add(obligation.getId());
+            index++;
+        }
+        if (!CollectionUtils.isEmpty(repeatedIds)) {
```

Comment:
Do we want to error out on repeated obligations?  If so, it would be better to define obligation ids as an ordered set to avoid duplication.

Reply:
yes, obligation id should be unique and with sets we cannot arbitrarily decide which one to keep as mentioned on the previous comment